lib/pure/parsecsv.nim

```diff
@@ -162,9 +173,52 @@ proc readRow*(my: var CsvParser, columns = 0): bool =
           $col & " columns")
   inc(my.currRow)
 
+proc tryReadRow*(my: var CsvParser, columns = 0): bool =
+  ## reads the next row without raising `CsvError`;
+  ## Returns false if the end of the file
+  ## has been encountered else true.
+  ##
+  ## Blank lines are skipped.
+  var col = 0 # current column
+  var oldpos = my.bufpos
+  while my.buf[my.bufpos] != '\0':
+    var oldlen = my.row.len
+    if oldlen < col+1:
+      setLen(my.row, col+1)
+      my.row[col] = ""
+    parseField(my, my.row[col])
+    inc(col)
+    if my.buf[my.bufpos] == my.sep:
+      inc(my.bufpos)
+    else:
+      case my.buf[my.bufpos]
+      of '\c', '\l':
+        # skip empty lines:
+        while true:
+          case my.buf[my.bufpos]
+          of '\c': my.bufpos = handleCR(my, my.bufpos)
+          of '\l': my.bufpos = handleLF(my, my.bufpos)
+          else: break
+      of '\0': discard
+      else: discard  # no delimeter - ignore
+      break
+
+  setLen(my.row, col)
+  result = col > 0
+  inc(my.currRow)
+
+proc readRow*(my: var CsvParser, filename: string,
+           separator = ',', quote = '"', escape = '\0',
+           skipInitialSpace = false, columns = 0): bool =
+  ## Opens the file, then call the generic `tryReadRow()`
+  if my.filename.isNil:
+    my.open(filename, separator, quote, escape, skipInitialSpace)
+  result = my.tryReadRow(columns = columns)
+
 proc close*(my: var CsvParser) {.inline.} =
   ## closes the parser `my` and its associated input stream.
   lexbase.close(my)
+  my.filename = nil
```

Comment:
Setting the filename to nil is a mechanism to identify that CsvParser isn't already in use (one alternative is to use a flag to identify this).

The real issue with exceptions is when reading mal-formed CSV files (rather than an exception for failing to open the file).  
Exceptions in the while loop test are difficult to recover from, so is helpful having an exception-free readRow().
An exception when opening the file is easy to handle.  It usually isn't part of a loop test construct.

-------------------
If I can guess at the difference in perspectives . . .

It is that you are assuming a csv file is a file of data with every line having the correct number of columns of data, and lack of uniformity is an exception (which is correct for this type of csv file)

I am coming from the perspective that this is one of many use-cases.  One of the other cases is to use parsecsv to split text files based on a character that may or may not exist on each line, so I expect that splitting a line by a character should give me an unsplit line (a seq of one element that is the whole line, or a seq of numerous elements).  However, parsecsv throws an exception when it doesn't find the separator character.  Hence, it would be helpful to have an exception-free readRow().

-------------------
The open() using a filename rather than a stream, is for making it helpful to open files based on the filename (save on the effort of handling a stream).
I assume this is a fairly common thing that will be done.  I didn't add an exception-free open(), but maybe that should be added too??